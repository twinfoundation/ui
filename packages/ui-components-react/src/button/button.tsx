// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
import { Button as FlowbiteButton } from "flowbite-react";
import { memo, useMemo, type JSX } from "react";
import type { ButtonProps } from "./buttonProps";

const BASE_CLASSES = "border-2 border-transparent focus:outline-none focus:ring-2";
const GHOST_BASE_CLASSES = "focus:outline-none focus:ring-1";

const colorClasses = {
	primary: `${BASE_CLASSES} text-white bg-surface-button dark:bg-surface-button hover:enabled:bg-surface-button-hover dark:hover:enabled:bg-surface-button-hover focus:ring-surface-button-pressed`,
	secondary: `${BASE_CLASSES} text-white bg-surface-button-alt dark:bg-surface-button-alt hover:enabled:bg-surface-button-alt-hover dark:hover:enabled:bg-surface-button-alt-hover focus:ring-surface-button-alt-pressed`,
	error: `${BASE_CLASSES} text-white bg-error dark:bg-error hover:enabled:bg-system-error-tints-600 dark:hover:enabled:bg-system-error-tints-600 focus:ring-system-error-tints-200`,
	warning: `${BASE_CLASSES} text-white bg-warning dark:bg-warning hover:enabled:bg-system-warning-tints-600 dark:hover:enabled:bg-system-warning-tints-600 focus:ring-system-warning-tints-200`,
	success: `${BASE_CLASSES} text-white bg-success dark:bg-success hover:enabled:bg-system-success-tints-600 dark:hover:enabled:bg-system-success-tints-600 focus:ring-system-success-tints-200`,
	info: `${BASE_CLASSES} text-white bg-information dark:bg-information hover:enabled:bg-system-information-tints-600 dark:hover:enabled:bg-system-information-tints-600 focus:ring-system-information-tints-200`,
	plain: `${BASE_CLASSES} text-black dark:text-invert bg-surface-second dark:bg-surface-third-dark hover:enabled:bg-surface-third focus:ring-gray-200 dark:hover:enabled:bg-surface-second-dark dark:focus:ring-surface-button-pressed`,
	ghost: `${GHOST_BASE_CLASSES} text-black dark:text-invert hover:enabled:bg-gray-50 focus:bg-gray-100 focus:ring-gray-200 dark:hover:enabled:bg-surface-second-dark dark:focus:ring-surface-button-pressed`,
	dark: `${BASE_CLASSES} text-white dark:text-black bg-gray-800 dark:bg-surface-second hover:enabled:bg-gray-700 focus:ring-gray-200 dark:hover:enabled:bg-gray-600 dark:focus:ring-gray-700`
} as const;

const buttonSizes = {
	xs: "!h-7 !w-7",
	sm: "!h-9 !w-9",
	md: "!h-11 !w-11",
	lg: "!h-14 !w-14",
	xl: "!h-16 !w-16"
} as const;

/**
 * Button icon size.
 */
export type IconSize = keyof typeof iconSizes;

const iconSizes = {
	xs: "h-3.5 w-3.5",
	sm: "h-4 w-4",
	md: "h-5 w-5",
	lg: "h-6 w-6",
	xl: "h-7 w-7"
} as const;

/**
 * Button component.
 */
export const Button = memo(
	({
		color = "primary",
		size = "md",
		outline = false,
		iconOnly = false,
		showButtonText = true,
		buttonText = "",
		showLeftIcon = true,
		leftIcon: LeftIcon,
		rightIcon: RightIcon,
		showRightIcon = true,
		icon: Icon,
		disabled = false,
		className,
		children,
		...rest
	}: ButtonProps): JSX.Element => {
		const iconSize = iconSizes[size as IconSize];

		const buttonContent = useMemo(
			() => (
				<div className="inline-flex items-center justify-center">
					{showLeftIcon && LeftIcon && (
						<div className="flex items-center">
							<LeftIcon className={`mr-1.5 ${iconSize}`} />
						</div>
					)}
					{showButtonText && (buttonText || children)}
					{showRightIcon && RightIcon && (
						<div className="flex items-center">
							<RightIcon className={`ml-1.5 ${iconSize}`} />
						</div>
					)}
				</div>
			),
			[
				children,
				LeftIcon,
				RightIcon,
				buttonText,
				iconSize,
				showButtonText,
				showLeftIcon,
				showRightIcon
			]
		);

		const iconContent = useMemo(
			() =>
				Icon && (
					<div className="flex h-full w-full items-center justify-center">
						<Icon className={iconSize} />
					</div>
				),
			[Icon, iconSize]
		);

		const iconOnlyClasses = useMemo(() => {
			if (iconOnly) {
				return `!aspect-square !rounded-full !p-0 ${buttonSizes[size as IconSize]} [&>span]:!p-0 [&>span]:!m-0`;
			}
			return "rounded-lg";
		}, [iconOnly, size]);

		const buttonColorClasses = useMemo(
			() => colorClasses[color as keyof typeof colorClasses],
			[color]
		);

		let ghostClasses = "";
		if (outline && color === "ghost") {
			ghostClasses = "border-[1px]";
		}

		return (
			<FlowbiteButton
				color={color}
				size={size}
				outline={outline}
				disabled={disabled}
				className={`${buttonColorClasses} ${iconOnlyClasses} ${className ?? ""} ${ghostClasses}`}
				{...rest}
			>
				{iconOnly ? iconContent : buttonContent}
			</FlowbiteButton>
		);
	}
);

Button.displayName = "Button";