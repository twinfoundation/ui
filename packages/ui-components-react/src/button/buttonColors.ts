// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.

/**
 * Button colors.
 */
export const ButtonColors = {
	/**
	 * Primary button - Used for primary actions
	 */
	Primary: "primary",

	/**
	 * Secondary button - Used for secondary actions
	 */
	Secondary: "secondary",

	/**
	 * Ghost button - Used for tertiary actions
	 */
	Ghost: "ghost",

	/**
	 * Success button - Used for success states
	 */
	Success: "success",

	/**
	 * Error button - Used for error states
	 */
	Error: "error",

	/**
	 * Warning button - Used for warning states
	 */
	Warning: "warning",

	/**
	 * Info button - Used for informational states
	 */
	Info: "info"
} as const;

/**
 * Type for button colors
 */
export type ButtonColor = (typeof ButtonColors)[keyof typeof ButtonColors];
