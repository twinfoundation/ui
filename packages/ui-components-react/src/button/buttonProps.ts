// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
import type { ButtonProps as FlowbiteButtonProps } from "flowbite-react";
import type { ElementType, ReactNode } from "react";
import type { ButtonColors } from "./buttonColors";
import type { ButtonSize } from "./buttonSizes";

/**
 * Button component props interface
 */
export interface ButtonProps extends Omit<FlowbiteButtonProps, "color" | "label"> {
	/**
	 * Children elements
	 */
	children?: ReactNode;

	/**
	 * The color variant of the button
	 */
	color?: ButtonColors;

	/**
	 * The size variant of the button
	 */
	size?: ButtonSize;

	/**
	 * Whether to show the button in outline style
	 */
	outline?: boolean;

	/**
	 * Whether to show only the icon without text
	 */
	iconOnly?: boolean;

	/**
	 * Whether to show the button text
	 */
	showButtonText?: boolean;

	/**
	 * The text content of the button
	 */
	buttonText?: string;

	/**
	 * Whether to show the left icon
	 */
	showLeftIcon?: boolean;

	/**
	 * The component type for the left icon
	 */
	leftIcon?: ElementType;

	/**
	 * The component type for the right icon
	 */
	rightIcon?: ElementType;

	/**
	 * Whether to show the right icon
	 */
	showRightIcon?: boolean;

	/**
	 * The component type for the icon (when using iconOnly)
	 */
	icon?: ElementType;

	/**
	 * Whether the button is disabled
	 */
	disabled?: boolean;

	/**
	 * Additional CSS class names
	 */
	className?: string;
}
