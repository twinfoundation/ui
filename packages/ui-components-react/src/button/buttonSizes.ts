// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.

/**
 * Button sizes.
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export const ButtonSizes = {
	/**
	 * Extra Small.
	 */
	ExtraSmall: "xs",

	/**
	 * Small.
	 */
	Small: "sm",

	/**
	 * Medium.
	 */
	Medium: "md",

	/**
	 * Large.
	 */
	Large: "lg",

	/**
	 * Extra Large.
	 */
	ExtraLarge: "xl"
} as const;

/**
 * Button sizes type.
 */
export type ButtonSize = (typeof ButtonSizes)[keyof typeof ButtonSizes];
