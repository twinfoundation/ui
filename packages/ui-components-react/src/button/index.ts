// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.

export { Button } from "./button";
export type * from "./buttonColors";
export type * from "./buttonProps";
export type * from "./buttonSizes";
