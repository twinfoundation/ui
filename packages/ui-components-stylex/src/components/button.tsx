// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
import * as stylex from "@stylexjs/stylex";
import { memo, useMemo, type JSX } from "react";
import type { ButtonProps } from "./buttonProps";

const colorStyles = stylex.create({
	primary: {
		color: "white",
		backgroundColor: "var(--surface-button)",
		":hover:enabled": { backgroundColor: "var(--surface-button-hover)" },
		":focus": { outline: "2px solid var(--surface-button-pressed)" },
	},
	secondary: {
		color: "white",
		backgroundColor: "var(--surface-button-alt)",
		":hover:enabled": { backgroundColor: "var(--surface-button-alt-hover)" },
		":focus": { outline: "2px solid var(--surface-button-alt-pressed)" },
	},
	error: {
		color: "white",
		backgroundColor: "var(--error)",
		":hover:enabled": { backgroundColor: "var(--system-error-tints-600)" },
		":focus": { outline: "2px solid var(--system-error-tints-200)" },
	},
	warning: {
		color: "white",
		backgroundColor: "var(--warning)",
		":hover:enabled": { backgroundColor: "var(--system-warning-tints-600)" },
		":focus": { outline: "2px solid var(--system-warning-tints-200)" },
	},
	success: {
		color: "white",
		backgroundColor: "var(--success)",
		":hover:enabled": { backgroundColor: "var(--system-success-tints-600)" },
		":focus": { outline: "2px solid var(--system-success-tints-200)" },
	},
	info: {
		color: "white",
		backgroundColor: "var(--information)",
		":hover:enabled": { backgroundColor: "var(--system-information-tints-600)" },
		":focus": { outline: "2px solid var(--system-information-tints-200)" },
	},
	plain: {
		color: "black",
		backgroundColor: "var(--surface-second)",
		":hover:enabled": { backgroundColor: "var(--surface-third)" },
		":focus": { outline: "2px solid #e5e7eb" },
		"@media (prefers-color-scheme: dark)": {
			color: "var(--invert)",
			backgroundColor: "var(--surface-third-dark)",
			":hover:enabled": { backgroundColor: "var(--surface-second-dark)" },
			":focus": { outline: "2px solid var(--surface-button-pressed)" },
		},
	},
	ghost: {
		color: "black",
		backgroundColor: "transparent",
		":hover:enabled": { backgroundColor: "#f9fafb" },
		":focus": { outline: "1px solid #e5e7eb", backgroundColor: "#f3f4f6" },
		"@media (prefers-color-scheme: dark)": {
			color: "var(--invert)",
			":hover:enabled": { backgroundColor: "var(--surface-second-dark)" },
			":focus": { outline: "1px solid var(--surface-button-pressed)" },
		},
	},
	dark: {
		color: "white",
		backgroundColor: "#1f2937",
		":hover:enabled": { backgroundColor: "#374151" },
		":focus": { outline: "2px solid #e5e7eb" },
		"@media (prefers-color-scheme: dark)": {
			color: "black",
			backgroundColor: "var(--surface-second)",
			":hover:enabled": { backgroundColor: "#4b5563" },
			":focus": { outline: "2px solid #374151" },
		},
	},
});

const baseStyle = stylex.create({
	base: {
		display: "inline-flex",
		alignItems: "center",
		justifyContent: "center",
		borderWidth: "2px",
		borderStyle: "solid",
		borderColor: "transparent",
		fontWeight: 500,
		textDecoration: "none",
		cursor: "pointer",
		transition: "all 0.2s ease-in-out",
		outline: "none",
		userSelect: "none",
		":disabled": {
			opacity: 0.6,
			cursor: "not-allowed"
		},
	},
	rounded: {
		borderRadius: "0.5rem"
	},
	iconOnly: {
		aspectRatio: "1",
		borderRadius: "9999px",
		padding: 0,
	},
	ghostOutline: {
		borderWidth: "1px",
		borderStyle: "solid",
		borderColor: "currentColor"
	},
});

const sizeStyles = stylex.create({
	xs: { height: "1.75rem", width: "1.75rem", padding: "0.25rem", fontSize: "0.75rem" },
	sm: { height: "2.25rem", width: "2.25rem", padding: "0.375rem 0.75rem", fontSize: "0.875rem" },
	md: { height: "2.75rem", width: "2.75rem", padding: "0.5rem 1rem", fontSize: "1rem" },
	lg: { height: "3.5rem", width: "3.5rem", padding: "0.75rem 1.5rem", fontSize: "1.125rem" },
	xl: { height: "4rem", width: "4rem", padding: "1rem 2rem", fontSize: "1.25rem" },
});

const iconSizeStyles = stylex.create({
	xs: { height: "0.875rem", width: "0.875rem" },
	sm: { height: "1rem", width: "1rem" },
	md: { height: "1.25rem", width: "1.25rem" },
	lg: { height: "1.5rem", width: "1.5rem" },
	xl: { height: "1.75rem", width: "1.75rem" },
});

export type IconSize = keyof typeof iconSizeStyles;

export const Button = memo(({
	color = "primary",
	size = "md",
	outline = false,
	iconOnly = false,
	showButtonText = true,
	buttonText = "",
	showLeftIcon = true,
	leftIcon: LeftIcon,
	rightIcon: RightIcon,
	showRightIcon = true,
	icon: Icon,
	disabled = false,
	className,
	children,
	...rest
}: ButtonProps): JSX.Element => {
	const iconSize = iconSizeStyles[size as IconSize];

	const buttonContent = useMemo(
		() => (
			<div style={{ display: "inline-flex", alignItems: "center", justifyContent: "center" }}>
				{showLeftIcon && LeftIcon && (
					<span style={{ display: "flex", alignItems: "center", marginRight: "0.375rem" }}>
						<LeftIcon {...stylex.props(iconSize)} />
					</span>
				)}
				{showButtonText && (buttonText || children)}
				{showRightIcon && RightIcon && (
					<span style={{ display: "flex", alignItems: "center", marginLeft: "0.375rem" }}>
						<RightIcon {...stylex.props(iconSize)} />
					</span>
				)}
			</div>
		),
		[children, LeftIcon, RightIcon, buttonText, iconSize, showButtonText, showLeftIcon, showRightIcon]
	);

	const iconContent = useMemo(
		() =>
			Icon && (
				<span style={{ display: "flex", height: "100%", width: "100%", alignItems: "center", justifyContent: "center" }}>
					<Icon {...stylex.props(iconSize)} />
				</span>
			),
		[Icon, iconSize]
	);

	const buttonStyles = [
		baseStyle.base,
		iconOnly ? baseStyle.iconOnly : baseStyle.rounded,
		sizeStyles[size],
		colorStyles[color],
		iconOnly && sizeStyles[size],
		outline && color === "ghost" ? baseStyle.ghostOutline : null,
	];

	return (
		<button
			type="button"
			{...stylex.props(buttonStyles)}
			disabled={disabled}
			className={className}
			{...rest}
		>
			{iconOnly ? iconContent : buttonContent}
		</button>
	);
});

Button.displayName = "Button";
