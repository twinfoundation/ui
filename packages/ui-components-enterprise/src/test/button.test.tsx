// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, it, expect, vi } from 'vitest';
import { Button } from '../button/button';

describe('Button Component', () => {
  it('renders with default props', () => {
    render(<Button>Click me</Button>);
    const button = screen.getByRole('button', { name: /click me/i });
    expect(button).toBeInTheDocument();
  });

  it('renders with custom text', () => {
    render(<Button>Custom Button Text</Button>);
    const button = screen.getByRole('button', { name: /custom button text/i });
    expect(button).toBeInTheDocument();
  });

  it('applies primary color by default', () => {
    render(<Button>Primary Button</Button>);
    const button = screen.getByRole('button');
    // Check if the button has the primary color classes
    expect(button).toHaveClass('bg-surface-button');
  });

  it('handles click events', async () => {
    const handleClick = vi.fn();
    const user = userEvent.setup();
    
    render(<Button onClick={handleClick}>Click me</Button>);
    const button = screen.getByRole('button');
    
    await user.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('can be disabled', () => {
    render(<Button disabled>Disabled Button</Button>);
    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
  });

  it('applies custom className', () => {
    render(<Button className="custom-class">Button</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass('custom-class');
  });

  it('forwards additional props', () => {
    render(<Button data-testid="custom-button" aria-label="Custom label">Button</Button>);
    const button = screen.getByTestId('custom-button');
    expect(button).toHaveAttribute('aria-label', 'Custom label');
  });

  it('renders with different colors', () => {
    const { rerender } = render(<Button color="error">Error Button</Button>);
    let button = screen.getByRole('button');
    expect(button).toHaveClass('bg-error');

    rerender(<Button color="success">Success Button</Button>);
    button = screen.getByRole('button');
    expect(button).toHaveClass('bg-success');

    rerender(<Button color="warning">Warning Button</Button>);
    button = screen.getByRole('button');
    expect(button).toHaveClass('bg-warning');
  });

  it('renders with different sizes', () => {
    const { rerender } = render(<Button size="xs" iconOnly>Small Button</Button>);
    let button = screen.getByRole('button');
    expect(button).toHaveClass('!h-7');

    rerender(<Button size="lg" iconOnly>Large Button</Button>);
    button = screen.getByRole('button');
    expect(button).toHaveClass('!h-14');
  });

  it('renders as icon only', () => {
    render(<Button iconOnly>Icon Button</Button>);
    const button = screen.getByRole('button');
    expect(button).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<Button>Snapshot Test</Button>);
    expect(container).toMatchSnapshot();
  });
}); 