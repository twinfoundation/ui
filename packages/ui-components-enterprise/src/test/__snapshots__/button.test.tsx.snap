// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > matches snapshot 1`] = `
<div>
  <button
    class="group relative flex items-stretch justify-center p-0.5 text-center font-medium transition-[color,background-color,border-color,text-decoration-color,fill,stroke,box-shadow] focus:z-10 border-2 border-transparent focus:outline-none focus:ring-2 text-white bg-surface-button dark:bg-surface-button hover:enabled:bg-surface-button-hover dark:hover:enabled:bg-surface-button-hover focus:ring-surface-button-pressed rounded-lg"
    type="button"
  >
    <span
      class="flex items-stretch transition-all duration-200 rounded-md px-4 py-2 text-sm"
    >
      <div
        class="inline-flex items-center justify-center"
      >
        Snapshot Test
      </div>
    </span>
  </button>
</div>
`;
